[#161647315] Enable self service postgres extensions

AP-Hunt opened the pull request:
What
----
Enables self-servicing of supported Postgres extensions (specified in config). Uses AWS RDS parameter group names to inform which parameter group name an instance should be created in.

How to review
-------------
Code review
See related [PR](https://github.com/alphagov/paas-cf/pull/1745) on [paas-cf](https://github.com/alphagov/paas-cf/pulls)

Who can review
--------------
Not @mogds or @AP-Hunt or @LeePorte

AP-Hunt committed:
Resolve test compilation failures

Tests in the awsrds package were failing to compile due to unused variables
under both Go 1.8 and 1.11.

AP-Hunt committed:
Allow self-service provisioning of supported Postgres extensions

Users have requested that Postgres extensions be enabled on existing databases
in the past, which has required manual intervention. In this commit, users can
self-serve supported extensions for Postgresql (specified in configuration, per
plan) when creating the service e.g.

    cf create-service postgres tiny-unencrypted-9.5 \
      -c '{"enabled_extensions": ["postgis", "pg_stat_statements"]}'

Some extensions require shared preload libraries to be enabled in the AWS RDS
parameter group. To support this, parameter group names are supplied in the
configuration file, and decoded to determine which special case extensions they
support. E.g.

    rdsbroker-postgres95-envname-pg-stat-statements

The parameter group names are used in conjuction with the requested extensions
to determine which parameter group the RDS instance should be created in.

AP-Hunt committed:
Prevent db parameter group changing during an update

The DB parameter group can only be decided at creation time. When updating, it
should be use the same value as the existing db instance.

AP-Hunt committed:
Enable extensions by default in parameter groups

Before users were able to enable extensions for themselves, a set of extensions
were enabled on every parameter group by default. This commit restores the
default extensions behaviour in order to maintain backwards compatability.

Users are still able to self-serve any supported extensions, but the default
extensions for that plan will also be enabled.

AP-Hunt committed:
Parameter group names expect to use the configurable db name prefix

Databases are created with a configurable prefix. Provided parameter groups are
now expected to have a matching prefix in order to be correclty decoded. E.g.

     <db_prefix>-postgres10-envname

Integration tests now create and destroy their own parameter groups using the
prefix

AP-Hunt committed:
Increase the threshold for marking integration tests as slow

The default value is 5 seconds, thus all tests are marked as slow. Increasing
the timeout to 15 minutes (900 seconds) should make the slow tests marking more
useful.

AP-Hunt committed:
Create RDS parameter group names on demand using an internal naming convention

Rather than have parameter groups passed in with structured, parseable names,
allow the RDS broker to create its own parameter groups. This change allows the
parameter group names to be more flexible, because they're not required to be
parsed. It also reduces the coupling between paas-cf and paas-rds-broker,
because they no longer need to agree on a naming convention.

AP-Hunt committed:
Bump gomega and ginkgo versions

Gomega => 1.4.3
Ginkgo => 1.7.0

Also bumps container version the latest tag with the latest version of ginkgo

governmentpaas/cf-acceptance-tests container => b77e27029dfcb85f6c58f0a59298d05cc5eeb903

AP-Hunt committed:
Enable stream flag on integration test invocation

AP-Hunt committed:
Add `defer GinkgoRecover()` to provision manager go routine

During integration tests, we run some code inside a goroutine. This could panic
(e.g. via assertions), and Ginkgo was unable to recover from the panic and
provide useful reports. Adding `defer GinkgoRecover()` to the top of the
goroutines allows Ginkgo to recover.

AP-Hunt committed:
Set parameter group on databases being restored from a snapshot

If no parameter group is specified, the default one is used. Setting it
required an upgrade of the aws-sdk-go package.

AP-Hunt committed:
Cleanup potentially left-over databases after integration tests

If integration tests fail for some reason, they can leave behind database
instances that would otherwise never be terminated. Post-test-suite cleanup now
checks for any databases with the right prefix that aren't in the "deleting"
state, and deleted them.

AP-Hunt committed:
Update documentation to mention extensions and how to configure them.

Mentions `enabled_extensions` provision parameter in README, and adds
configuration parameters to configuration docs and sample.

bandesz commented on README.md:
```diff
@@ -179,6 +182,15 @@ It is noted that masterPassword and binding username are deterministically gener
+to be enabled. In RDS, preload libraries are specified as a parameter in the parameter group, and are enabled for all 
+databases in that group.
+
+The RDS broker will created new parameter groups to support new combinations of extensions as necessary, or otherwise 
```
Typo: created -> create

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -103,6 +103,80 @@ func DestroySecurityGroup(id *string, session *session.Session) error {
+		func(page *rds.DescribeDBInstancesOutput, lastPage bool) bool {
+			if len(page.DBInstances) > 0 {
+				for _, instance := range page.DBInstances {
+					// Only include databases with the correct prefix
```
these comments can be removed, as both are stating the obvious.

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -103,6 +103,80 @@ func DestroySecurityGroup(id *string, session *session.Session) error {
+				for _, instance := range page.DBInstances {
+					// Only include databases with the correct prefix
+					if strings.HasPrefix(aws.StringValue(instance.DBInstanceIdentifier), prefix) {
+						// Only databases in the deleting state
```
remove comment

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -103,6 +103,80 @@ func DestroySecurityGroup(id *string, session *session.Session) error {
+
+	for _, instance := range requiringDeletion {
+		_, err := rdsSvc.DeleteDBInstance(&rds.DeleteDBInstanceInput{
+			DBInstanceIdentifier: aws.String(instance),
```
I suspect we should set `SkipFinalSnapshot` to `true`, as we don't want to have a final snapshot of a test database.

bandesz commented on ci/blackbox/integration_suite_test.go:
```diff
@@ -90,12 +92,20 @@ func TestSuite(t *testing.T) {
 				Region: aws.String(suiteData.RdsBrokerConfig.RDSConfig.Region)},
 			)
+
+			err := CleanUpTestDatabaseInstances(suiteData.RdsBrokerConfig.RDSConfig.DBPrefix, awsSession)
```
I really don't like these cleanups as a later code change can accidentally set the DBPrefix to an empty string and a test run will wipe out all databases in dev. What is the reason we can't do an explicit cleanup for any databases created?

AP-Hunt replied:
The rationale here was delete databases which tests didn't tidy up themselves, for example because the test failed before it could get to that point. I agree with your subsequent comment though.

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -103,6 +103,80 @@ func DestroySecurityGroup(id *string, session *session.Session) error {
+}
+
+func CleanUpTestDatabaseInstances(prefix string, awsSession *session.Session) error {
+	rdsSvc := rds.New(awsSession)
```
if we want to keep these cleanups we should make sure the prefix starts with `build-test-`, otherwise panic.

AP-Hunt replied:
Yep, agree here. Probably a good idea.

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -103,6 +103,80 @@ func DestroySecurityGroup(id *string, session *session.Session) error {
 }
 
+func CleanUpParameterGroups(prefix string, session *session.Session) error {
+	rdsService := rds.New(session)
```
if we want to keep these cleanups we should make sure the prefix starts with `build-test-`, otherwise panic.

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,170 @@
+	return strings.HasPrefix(err.Error(), rds.ErrCodeDBParameterGroupNotFoundFault)
+}
+
+func normalise(value string) string {
```
this name is too generic

bandesz commented on rdsbroker/broker.go:
```diff
@@ -546,17 +578,58 @@ func (b *RDSBroker) LastOperation(
+		return false
+	}
+
+	if parameters.Extensions == nil {
```
this if block is unnecessary, Go handles nil or empty list the same way if the type is list.

bandesz commented on rdsbroker/broker.go:
```diff
@@ -546,17 +578,58 @@ func (b *RDSBroker) LastOperation(
+	extensions := parameters.Extensions
+
+	supported := plan.RDSProperties.AllowedExtensions
+	supportedSlice := []string{}
```
supportedSlice is unused

bandesz commented on rdsbroker/broker.go:
```diff
@@ -546,17 +578,58 @@ func (b *RDSBroker) LastOperation(
+	return true, ""
+}
+
+func extensionIsSupported(extensions []*string, s string) bool {
```
we could inline this in the previous function so it doesn't clutter the global function space.

AP-Hunt replied:
I think the flow of control is easier to follow with this method extracted. Unless clutter in the global function space is a major problem, I feel readability is more important here.

bandesz commented on rdsbroker/broker.go:
```diff
@@ -591,7 +661,13 @@ func (b *RDSBroker) updateDBSettings(instanceID string, dbInstance *rds.DBInstan
 	}
 
-	modifyDBInstanceInput := b.newModifyDBInstanceInput(instanceID, servicePlan)
+	existingInstance, err := b.dbInstance.Describe(b.dbInstanceIdentifier(instanceID))
```
why do we need to load the database details again? It's confusing as it was loaded previously into `dbInstance`.

bandesz commented on rdsbroker/broker.go:
```diff
@@ -770,6 +851,11 @@ func (b *RDSBroker) createDBInstance(instanceID string, servicePlan ServicePlan,
 
+	parameterGroupName, err := b.parameterGroupsSelector.SelectParameterGroup(servicePlan, provisionParameters)
+	if err != nil {
+		return &rds.CreateDBInstanceInput{}, err
```
if there was an error, then return with nil.

bandesz commented on rdsbroker/broker.go:
```diff
@@ -865,7 +957,7 @@ func (b *RDSBroker) newModifyDBInstanceInput(instanceID string, servicePlan Serv
 }
 
-func (b *RDSBroker) dbTags(action, serviceID, planID, organizationID, spaceID, skipFinalSnapshot, originSnapshotIdentifier string) map[string]string {
+func (b *RDSBroker) dbTags(action, serviceID, planID, organizationID, spaceID, skipFinalSnapshot, originSnapshotIdentifier string, extensions []string) map[string]string {
```
This method has too many parameters now and it is really easy to screw up to pass a wrong value for a parameter. Please create a struct containing all tags.

bandesz commented on rdsbroker/broker.go:
```diff
@@ -902,5 +994,9 @@ func (b *RDSBroker) dbTags(action, serviceID, planID, organizationID, spaceID, s
 	}
+
+	if len(extensions) > 0 {
+		tags[awsrds.TagExtensions] = strings.Join(extensions, ":")
```
Why joining by `:`, it's seems like an unnatural choice. Anything against comma?

AP-Hunt replied:
It was initially comma-separated, but then we discovered that tag values can't contain commas. A colon seemed like a reasonable replacement.

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_Tagging.html#Overview.Tagging

bandesz commented on rdsbroker/broker_test.go:
```diff
@@ -541,14 +573,21 @@ var _ = Describe("RDS Broker", func() {
 			})
 
 			It("sets the right tags", func() {
+				jsondata := []byte(`{"enabled_extensions": ["postgis", "pg_stat_statements"]}`)
```
this should be jsonData, the rawparams variable seems unnecessary (and in two other places)

bandesz commented on rdsbroker/broker_test.go:
```diff
@@ -1182,6 +1287,68 @@ var _ = Describe("RDS Broker", func() {
 			}
 
+			existingDbInstance = &rds.DBInstance{
+				AllocatedStorage:        nil,
```
It's completely unnecessary to set any nil params to nil, as that's the default value when you create a new object.

bandesz commented on rdsbroker/broker_test.go:
```diff
@@ -2240,6 +2399,7 @@ var _ = Describe("RDS Broker", func() {
 				{Key: aws.String("Created by"), Value: aws.String("AWS RDS Service Broker")},
 				{Key: aws.String("Service ID"), Value: aws.String("Service-1")},
 				{Key: aws.String("Plan ID"), Value: aws.String("Plan-1")},
+				{Key: aws.String("Extensions"), Value: aws.String("postgis,pg-stat-statements")},
```
here we separate the extensions with `,`, but in the code we use `:`.

AP-Hunt replied:
Good catch

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,170 @@
+	logger      lager.Logger
+}
+
+func NewParameterGroupSource(config Config, rdsInstance awsrds.RDSInstance, logger lager.Logger) ParameterGroupSource {
```
This should return with a pointer, as we have pointer receiver functions defined.

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,170 @@
+	return ParameterGroupSource{config, rdsInstance, logger}
+}
+
+func (groups *ParameterGroupSource) SelectParameterGroup(servicePlan ServicePlan, parameters ProvisionParameters) (string, error) {
```
it's convention to use only a short variable name like `p` (or e.g. `pgs`)

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,170 @@
+	return relevantExtensions
+}
+
+func isParameterGroupNotFoundError(err error) bool {
```
This way of checking the error is really brittle, the original AWS error will contain the rds.ErrCodeDBParameterGroupNotFoundFault code, so we shouldn't need to check if it's part of the error message.

AP-Hunt replied:
I actually agree with you here, however this is a requirement imposed by the [`awsrds.utils.HandleAWSError`](https://github.com/alphagov/paas-rds-broker/blob/enable_self_service_postgres_extensions_161647315/awsrds/utils.go#L67) function which is used to handle and return errors inside the `awsrds` module.

That function returns an error in the format "code: message" (except in the specific case of a database not found). That format is in turn expeced by other areas of the code base. 

I refactored that function, at one point, to match the [guidance](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/handling-errors.html) on how to handle errors with the Go SDK, but I found that more areas than I wanted to change as part of this work depended on it.

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,170 @@
+	return out
+}
+
+func rdsParameter(paramName string, paramValue string, applyMethod *string) *rds.Parameter {
```
Make the third parameter a string, if all other string params are already a string.

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			}
+
+			servicePlan = ServicePlan{
+				ID:          "",
```
You don't need to set the nil or empty string values, as that's the default when creating a new object.

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			}
+
+			provisionParameters = ProvisionParameters{
+				BackupRetentionPeriod:       0,
```
You don't need to set any of these values, as these are the default nil values.

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			Expect(name).To(ContainSubstring("101112"))
+		})
+
+		It("contains the broker name name", func() {
```
"name name" -> "name"

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			It("dash-separates extension names", func() {
+				provisionParameters.Extensions = []string{"pg_stat_statements", "pg_z"}
+
+				backup := SupportedPreloadExtensions["postgres10"]
```
This is a really bad code smell, when you have to backup and restore a global value.

bandesz replied:
Also this backup/restore doesn't work because you would need to create a copy of `SupportedPreloadExtensions["postgres10"]` (array and map variables are always references).

AP-Hunt replied:
The intent in this test was to have an extra extension that would only exist during the test, and would be extremely unlikely to be added as a legitimate extension in the future. However, that list is global and (meant to be) constant.

In this case, how would you suggest going about it?

bandesz replied:
I understand the intent, but as I said the backup/restore doesn't work. A test should never have any side effects.

The best solution would be to pass the global config as a parameter to the right instance. In the test you would set up your object with your test data instead.

AP-Hunt replied:
Makes sense. Maintain it as a `map[string][]DBExtension`, or refactor it to a struct?

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			}
+
+			provisionDetails = ProvisionParameters{
+				BackupRetentionPeriod:       0,
```
no need for any of the param values.

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+			servicePlan = ServicePlan{
+				ID:          "test-1",
+				Name:        "Test",
+				Description: "",
```
remove the empty param values.

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+
+			It("does not attempt to create the group", func() {
+				parameterGroupSource.SelectParameterGroup(servicePlan, provisionDetails)
+				createGroupCalls := rdsFake.CreateParameterGroupCallCount()
```
don't create a separate variable for createGroupCalls as it's unnecessary and confusing

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+				parameterGroupSource.SelectParameterGroup(servicePlan, provisionDetails)
+
+				Expect(rdsFake.CreateParameterGroupCallCount()).To(Equal(1))
+				createGroupCall := rdsFake.CreateParameterGroupArgsForCall(0)
```
this variable name is wrong, it's the first parameter passed to the `CreateParameterGroup` call so e.g. `createParameterGroupInput` is the correct name

bandesz commented on rdsbroker/parameter_groups_source_test.go:
```diff
@@ -0,0 +1,294 @@
+
+			DescribeTable(
+				"it modifies the created parameter group",
+				func(paramName *string, paramValue *string, applyMethod *string) {
```
This should not be a table-type test as that's used for different input + expectation sets. In this case you would just want to list all Expect() calls for all properties.

bandesz reviewed (changes requested):
I left comments to be address and also the custom acceptance tests fail with the following error message:

```Server error, status code: 502, error code: 10001, message: Service broker error: InvalidParameterValue: ParameterGroupFamily postgres10.5 is not a valid parameter group family```

The broker is now over a 1000 lines and the tests are over 3000, so that's definitely a code smell (although it wasn't really caused by this story).

I'm not completely happy with the overall solution:
 * we don't support updating an existing service which is the request we used to get on support (at least a follow-up story should address this)
 * this solution won't scale well with more allowed extensions added in the future and also doesn't support custom parameters in a parameter group. I just wanted to note that we might need to rework the whole thing in the future.

AP-Hunt committed:
Introduce RDSInstanceTags struct

The `rdsbroker/broker.dbTags` function had gotten too many parameters.
Introduced a struct to tidy up the parameters and make them easier to manage.

AP-Hunt committed:
Add `engine_family` property to plan configuration

RDS engine family strings aren't always the engine and the version
concatenated, for example in the case of postgres 10.5. To keep things simple,
a new `engine_family` property has been added to plan configuration. Its value
should be the correct family name for that engine and version, as reported by
`aws rds describe-db-engine-versions`.

Documentation and config samples are updated to include the new property.

bandesz commented on ci/helpers/aws_helper.go:
```diff
@@ -139,6 +139,10 @@ func CleanUpParameterGroups(prefix string, session *session.Session) error {
+		panic("Trying to clean up databases without the 'build-test-' prefix will fail")
+	}
+
 	rdsSvc := rds.New(awsSession)
```
This check should be added to the CleanUpParameterGroups function too.

AP-Hunt replied:
Good catch

bandesz commented on rdsbroker/broker.go:
```diff
@@ -853,7 +865,18 @@ func (b *RDSBroker) createDBInstance(instanceID string, servicePlan ServicePlan,
+		OrganizationID:           details.OrganizationGUID,
+		SpaceID:                  details.SpaceGUID,
+		SkipFinalSnapshot:        skipFinalSnapshotStr,
+		OriginSnapshotIdentifier: "",
```
This line is unnecessary, but I don't mind if it stays here.

AP-Hunt replied:
Oops, yeah, missed that one! I'll tidy it up.

AP-Hunt committed:
Do not set shared_preload_library property when it would be empty

When modifying a parameter group for a postgres database, we set the
`shared_preload_libraries` parameter when an extension that requires one is
requested. If none were required, we passed the empty string.

However, the RDS API will not accept an empty value for this parameter,
returning the below message. To avoid this, we now don't set the parameter if
no preload libraries are required.

    InvalidParameterValue: Empty value found in for parameter
    shared_preload_libraries

AP-Hunt committed:
Inject supported preload extensions

Maintains the global definition of supported preload extensions, but injects it
in to ParameterGroupsSource.

bandesz commented on rdsbroker/parameter_groups_source.go:
```diff
@@ -0,0 +1,166 @@
+		extensionsLogKey:  parameters.Extensions,
+	})
+
+	groupName := composeGroupName(pgs.config, servicePlan, parameters, map[string][]DBExtension{})
```
I suspect this should have `pgs. supportedPreloadExtensions` as the last parameter

AP-Hunt committed:
Maintain extensions enabled on a snapshot when restoring from one.

When restoring a postgres database from a snapshot, we enable any extensions
that were specified in the `Extensions` tag. In order to maintain backwards
compatibility with the snapshot and plan, we don't check whether the extensions
it specifies are supported by the plan.

bandesz commented on rdsbroker/broker_test.go:
```diff
@@ -560,6 +560,21 @@ var _ = Describe("RDS Broker", func() {
 				})
 			})
 
+			Context("when the snapshot had extensions enabled", func() {
```
This test should check whether we merge the extensions lists

bandesz replied:
I added an extra test case for this.

bandesz committed:
Remove unused masterConnectionString variable

bandesz committed:
Fix wrong sprintf placeholder
